### PR TITLE
PMs

### PR DESCRIPTION
**Pull request by nick-prat**
Added PMs to DUGRuleAction, DigWebAgent, SugarWebAgent, CapabilityPayloadDAO

### CODE REVIEW

File: src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js [hunk 1]
```diff
@@ -238,7 +243,16 @@ foam.CLASS({
         List grantPath = ((CrunchService) x.get("crunchService")).getGrantPath(x, receivingCapPayload.getId());
         processCapabilityList(x, grantPath, capabilityDataObjects, currentCapabilityDataObjects);
 
-        return find_(x, receivingCapPayload.getId());
+        var ret =  find_(x, receivingCapPayload.getId());
+        pm.setEndTime(System.currentTimeMillis());
+        return ret;
+      } catch(Throwable t) {
+        pm.setErrorMessage(t.getMessage());
```
Comment (jlhughes) [inline src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: Why not just call pm.error(x, t.getMessage()); ?

File: src/foam/nanos/dig/DUGRuleAction.js [hunk 2]
```diff
@@ -44,19 +44,28 @@ foam.CLASS({
...
-              getBearerToken(), 
-              getFormat(),
-              new foam.lib.AndPropertyPredicate(x, 
-                new foam.lib.PropertyPredicate[] {
-                  new foam.lib.ExternalPropertyPredicate(),
-                  new foam.lib.NetworkPropertyPredicate(), 
-                  new foam.lib.PermissionedPropertyPredicate()}),
-              true);
-              
+            var pm = new PM(DUGRuleAction.getOwnClassInfo().getId(), "dugrule." + dugRule.getId() + "." + dugRule.getName());
```
Comment (jlhughes) [inline src/foam/nanos/dig/DUGRuleAction.js]: No need to do this here, 
```suggestion
            var pm = new PM(DUGRuleAction.getOwnClassInfo().getId(), "dugrule.", dugRule.getId(), ".", dugRule.getName());
```
And PMs already combine names with ':' like a path.

File: src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js [hunk 3]
```diff
@@ -228,6 +229,10 @@ foam.CLASS({
     {
       name: 'put_',
       javaCode: `
+      var pm = new PM(CapabilityPayloadDAO.getOwnClassInfo().getId(), "capabilityPayloads.put");
+      pm.setStartTime(System.currentTimeMillis());
```
Comment (jlhughes) [inline src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: This is also not necessary. PM creation does this for you.

File: src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js [hunk 4]
```diff
@@ -238,7 +243,16 @@ foam.CLASS({
         List grantPath = ((CrunchService) x.get("crunchService")).getGrantPath(x, receivingCapPayload.getId());
         processCapabilityList(x, grantPath, capabilityDataObjects, currentCapabilityDataObjects);
 
-        return find_(x, receivingCapPayload.getId());
+        var ret =  find_(x, receivingCapPayload.getId());
+        pm.setEndTime(System.currentTimeMillis());
```
Comment (jlhughes) [inline src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: This is also not necessary as pm.error and pm.log both do this for you.

File: src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js [hunk 5]
```diff
@@ -228,6 +229,10 @@ foam.CLASS({
     {
       name: 'put_',
       javaCode: `
+      var pm = new PM(CapabilityPayloadDAO.getOwnClassInfo().getId(), "capabilityPayloads.put");
```
Comment (jlhughes) [inline src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: With regard to naming, in this case the first part is already CapabilityPayloadDAO, so the second 'capabilityPayloads.put can be just 'put'

File: src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js (hunk 1 above)
Comment (nick-prat) [reply src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: I saw that the `try { ... } catch (...) { ... } finally { ... pm.log(x) }` pattern was used elsewhere so I just stuck to that.

Comment (jlhughes) [reply src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js]: yes, the try/catch/finally is fine, 
It's the unnecessary calls to setErrorMessage, setIsError, setTime... 
just call 
pm.error(x, errorMesage) 
and
pm.log(x)

**Review by jlhughes — APPROVED** [review]

**Pull request merged**